File: dotnet/Base/GameX/App/ExportManager.cs
using GameX.Formats;
using GameX.Formats.Unknown;
using OpenStack.Vfx;
using System;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace GameX.App;

public static class ExportManager {
    const int MaxDegreeOfParallelism = 1; //4;

    public static async Task ExportAsync(Family family, Resource res, string filePath, Func<string, bool> match, int from, object option) {
        var fo = option as FileOption? ?? FileOption.Default;
        using var pak = family.OpenPakFile(res);
        // single
        if (pak is not MultiPakFile multi) { await ExportPakAsync(filePath, match, from, option, pak); return; }
        // write paks
        if ((fo & FileOption.Marker) != 0) {
            if (!string.IsNullOrEmpty(filePath) && !Directory.Exists(filePath)) Directory.CreateDirectory(filePath);
            var setPath = Path.Combine(filePath, ".set");
            using var w = new BinaryWriter(new FileStream(setPath, FileMode.Create, FileAccess.Write));
            await PakBinary.Stream.Write(new StreamPakFile(NetworkHost.Factory, new PakState(null, null, null, "Root")) {
                Files = [.. multi.PakFiles.Select(x => new FileSource { Path = x.Name })]
            }, w, "Set");
        }
        // multi
        foreach (var _ in multi.PakFiles) await ExportPakAsync(filePath, match, from, option, _);
    }

    static async Task ExportPakAsync(string filePath, Func<string, bool> match, int from, object option, PakFile _) {
        if (_ is not BinaryPakFile pak) throw new InvalidOperationException("s not a BinaryPakFile");
        var newPath = filePath != null ? Path.Combine(filePath, Path.GetFileName(pak.PakPath)) : null;
        // write pak
        await ExportPak2Async(pak, newPath, match, from, option,
            (file, idx) => { if ((idx % 50) == 0) Console.WriteLine($"{idx,6}> {file.Path}"); },
            (file, msg) => Console.WriteLine($"ERROR: {msg} - {file?.Path}"));
    }

    static async Task ExportPak2Async(BinaryPakFile source, string filePath, Func<string, bool> match, int from, object option, Action<FileSource, int> next, Action<FileSource, string> error) {
        var fo = option as FileOption? ?? FileOption.Default;
        source.Open();
        // create directory
        if (!string.IsNullOrEmpty(filePath) && !Directory.Exists(filePath)) Directory.CreateDirectory(filePath);
        // write files
        Parallel.For(from, source.Files.Count, new ParallelOptions { MaxDegreeOfParallelism = MaxDegreeOfParallelism }, async index => {
            var file = source.Files[index];
            if (match != null && !match(file.Path)) return;
            var newPath = filePath != null ? Path.Combine(filePath, file.Path) : null;

            // create directory
            var directory = newPath != null ? Path.GetDirectoryName(newPath) : null;
            if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory)) Directory.CreateDirectory(directory);

            // recursive extract pak, and exit
            if (file.Pak != null) { await ExportPak2Async(file.Pak, newPath, match, 0, option, next, error); return; }

            // ensure cached object factory
            if ((fo & FileOption.Object) != 0) source.EnsureCachedObjectFactory(file);

            // extract file
            try {
                await ExportFileAsync(file, source, newPath, option);
                if (file.Parts != null && (fo & FileOption.Raw) != 0)
                    foreach (var part in file.Parts) await ExportFileAsync(part, source, Path.Combine(filePath, part.Path), option);
                next?.Invoke(file, index);
            }
            catch (Exception e) { error?.Invoke(file, $"Exception: {e.Message}"); }
        });
        // write pak-raw
        if ((fo & FileOption.Marker) != 0) await new StreamPakFile(source, new PakState(source.Vfx, source.Game, source.Edition, filePath)).Write(null);
    }

    static async Task ExportFileAsync(FileSource file, BinaryPakFile source, string newPath, object option) {
        var fo = option as FileOption? ?? FileOption.Default;
        if (file.FileSize == 0 && file.PackedSize == 0) return;
        var oo = (file.CachedObjectOption as FileOption?) ?? FileOption.Default;
        if (file.CachedObjectOption != null && (fo & oo) != 0) {
            if (oo.HasFlag(FileOption.UnknownFileModel)) {
                var model = await source.LoadFileObject<IUnknownFileModel>(file, FamilyManager.UnknownPakFile);
                UnknownFileWriter.Factory("default", model).Write(newPath, false);
                return;
            }
            else if (oo.HasFlag(FileOption.BinaryObject)) {
                var obj = await source.LoadFileObject<object>(file);
                if (obj is IStream src) {
                    using var b2 = src.GetStream();
                    using var s2 = newPath != null
                        ? new FileStream(newPath, FileMode.Create, FileAccess.Write)
                        : (Stream)new MemoryStream();
                    b2.CopyTo(s2);
                    return;
                }
                PakBinary.HandleException(null, option, $"BinaryObject: {file.Path} @ {file.FileSize}");
                throw new InvalidOperationException();
            }
            else if (oo.HasFlag(FileOption.StreamObject)) {
                var obj = await source.LoadFileObject<object>(file);
                if (obj is IWriteToStream src) {
                    using var s2 = newPath != null
                        ? new FileStream(newPath, FileMode.Create, FileAccess.Write)
                        : (Stream)new MemoryStream();
                    src.WriteToStream(s2);
                    return;
                }
                PakBinary.HandleException(null, option, $"StreamObject: {file.Path} @ {file.FileSize}");
                throw new InvalidOperationException();
            }
        }
        using var b = await source.LoadFileData(file, option);
        using var s = newPath != null
            ? new FileStream(newPath, FileMode.Create, FileAccess.Write)
            : (Stream)new MemoryStream();
        b.CopyTo(s);
        if (file.Parts != null && (fo & FileOption.Raw) == 0)
            foreach (var part in file.Parts) {
                using var b2 = await source.LoadFileData(part, option);
                b2.CopyTo(s);
            }
    }
}

File: python/gamex/Base/app/exportManager.py
from __future__ import annotations
import os, io, asyncio, shutil
from openstk.poly import parallelFor, IStream, IWriteToStream
from gamex import FileOption, PakFile, MultiPakFile, BinaryPakFile

# ExportManager
class ExportManager:
    MaxDegreeOfParallelism = 1 #4

    @staticmethod
    async def exportAsync(family: Family, res: Resource, filePath: str, match: callable, from_: int, option: object) -> None:
        fo = option if isinstance(option, FileOption) else FileOption.Default
        with family.openPakFile(res) as pak:
            # single
            if not isinstance(pak, MultiPakFile): await ExportManager.exportPakAsync(filePath, match, from_, option, pak); return
            # write paks
            if FileOption.Marker in fo:
                if filePath and not os.path.isdir(filePath): os.makedirs(filePath)
                setPath = os.path.join(filePath, '.set')
            #     using var w = new BinaryWriter(new FileStream(setPath, FileMode.Create, FileAccess.Write));
            #     await PakBinary.Stream.Write(new StreamPakFile(NetworkHost.Factory, new PakState(null, null, null, "Root")) {
            #         Files = [.. multi.PakFiles.Select(x => new FileSource { Path = x.Name })]
            #     }, w, "Set");
            # multi
            for _ in pak.pakFiles: await ExportManager.exportPakAsync(filePath, match, from_, option, _)

    @staticmethod
    async def exportPakAsync(filePath: str, match: callable, from_: int, option: object, _: PakFile) -> None:
        pak = _
        if not isinstance(pak, BinaryPakFile): raise Exception('s not a BinaryPakFile')
        newPath = os.path.join(filePath, os.path.basename(pak.pakPath)) if filePath else None
        # write pak
        await ExportManager.exportPak2Async(pak, newPath, match, from_, option, \
            lambda file, idx: print(f'{idx:>6}> {file.path}') if (idx % 50) == 0 else None, \
            lambda file, msg: print(f'ERROR: {msg} - {file.path}'))

    @staticmethod
    async def exportPak2Async(source: BinaryPakFile, filePath: str, match: callable, from_: int, option: object, next: callable, error: callable) -> None:
        fo = option if isinstance(option, FileOption) else FileOption.Default
        # create directory
        if filePath and not os.path.isdir(filePath): os.makedirs(filePath)
        # write files
        async def lambdaX(index: int):
            file = source.files[index]
            # print(match(file.path))
            if match and not match(file.path): return
            newPath = os.path.join(filePath, file.path) if filePath else None
            # create directory
            directory = os.path.dirname(newPath) if newPath else None
            if directory and not os.path.isdir(directory): os.makedirs(directory)
            # recursive extract pak, and exit
            if file.pak: await exportPak2Async(file.pak, newPath, match, 0, option, next, error); return
            # ensure cached object factory
            if FileOption.Object in fo: source.ensureCachedObjectFactory(file)
            # extract file
            try:
                await ExportManager.exportFileAsync(file, source, newPath, option)
                if file.parts != None and FileOption.Raw in fo:
                    for part in file.parts: await ExportManager.exportFileAsync(part, source, os.path.join(filePath, part.path), option)
                if next: next(file, index)
            except Exception as e: error(file, repr(e)) if error else None # f'Exception: {str(e)}'
        await parallelFor(from_, len(source.files), { 'max': ExportManager.MaxDegreeOfParallelism }, lambdaX)
        # write pak-raw
        # if FileOption.Marker in fo: await StreamPakFile(source, new PakState(source.Vfx, source.Game, source.Edition, filePath)).Write(null)

    @staticmethod
    async def exportFileAsync(file: FileSource, source: BinaryPakFile, newPath: str, option: object) -> None:
        fo = option if isinstance(option, FileOption) else FileOption.Default
        if file.fileSize == 0 and file.packedSize == 0: return
        oo = file.cachedObjectOption if isinstance(file.cachedObjectOption, FileOption) else FileOption.Default
        if file.cachedObjectOption and bool(fo & oo):
            if FileOption.UnknownFileModel in oo:
                model = source.loadFileObject(IUnknownFileModel, file, FamilyManager.UnknownPakFile)
                # UnknownFileWriter.Factory('default', model).Write(newPath, false)
                return
            elif FileOption.BinaryObject in oo:
                obj = source.loadFileObject(object, file)
                if isinstance(obj, IStream):
                    with obj.getStream() as b2, open(newPath, 'wb') if newPath else io.BytesIO() as s2:
                        shutil.copyfileobj(b2, s2)
                    return
                PakBinary.handleException(None, option, f'BinaryObject: {file.Path} @ {file.FileSize}')
                raise Exception()
            elif FileOption.StreamObject in oo:
                obj = source.loadFileObject(object, file)
                if isinstance(obj, IWriteToStream):
                    with open(newPath, 'w') if newPath else io.BytesIO() as s2:
                        obj.writeToStream(s2)
                    return
                PakBinary.handleException(None, option, f'StreamObject: {file.Path} @ {file.FileSize}')
                raise Exception()
        with source.loadFileData(file, option) as b, open(newPath, 'wb') if newPath else io.BytesIO() as s:
            shutil.copyfileobj(b, s)
            if file.parts and FileOption.Raw not in fo:
                for part in file.parts:
                    with await source.loadFileData(part, option) as b2:
                        shutil.copyfileobj(b2, s)
